### ISSUE TITLE
Translation of taxonomy terms

### ISSUE THREAD

**Issue by amilenkov** [backdrop/backdrop-issues#4720]
I tried to find a similar issue, but I did not find it. It may be I am inexperienced, so if there is such a topic I apologize and please add my question to existent issue.

Backdrop now has the ability to specify the language of taxonomic terms. This is especially useful with the correct transliteration of URL address of the page of the corresponding taxonomy term in different languages.

A problem, however, is that when you are on a term for one language, switching languages ​​does not switch to translating the term to another language even though such translated term page exists.

As far as I understand, this is due to the fact that we can now set the language of the terms, but we cannot create translations and translation links between the translated terms.

There is also a problem with indexing pages of taxonomy terms in search engines. For example, you have a term in English, this page is indexed by Google and shown in search results, but when you visit the term page in English, the interface of the site is shown often in a different language. Not always, but very often. This seems to depend on the language in which the site was active before opening the link in the search results.

Is there a Drupal or Backdrop contributed module that accomplishes this?

In Drupal 7 i18n manages such problems correctly with a submodule.

I am not able to offer a solution, I am just reporting the problem.

**Comment by indigoxela** [comment]
> In Drupal 7 i18n manages such problems correctly with a submodule.

The module exists in [backdrop-contrib](https://github.com/backdrop-contrib/i18n) but has no stable release yet.

There are several open (but very likely outdated) issues. Testing i18n is very welcome!
Note: there's no upgrade path from D7 yet.

**Comment by amilenkov** [comment]
Thanks indigoxela,

I have known about https://github.com/backdrop-contrib/i18n for many years and follow the discussion as much as I can. And as far as I understand the Backdrop community seems to choose another approach to translation - for example Site name, Slogan are now successfully and very easy translated using the standard technique of User interface translation and not in the way that uses i18n with complex system of variables.

So I'm worried that installing the i18n won't conflict with the new translation techniques introduced in Backdrop. Which new techniques I like as simpler and easier to work with than working with i18n.

In addition, I have observed in Drupal 7 that i18n is a heavy and sometimes dangerous module, in the sense that it can not be completely uninstalled and once activated, some of its tables remain in the database, even if uninstalled and i18n permanently changes the operation of parts of the language system maintenance and translation.

I will test i18n on a non-essential test site, but I would be happy if the problem with the translating connections of the terms in the taxonomy is solved in the future in the way conceived by Backdrop, in the core itself, without the need for additional modules.

In fact, for me, apart from the problem of translating terms in the taxonomy described above, Backdrop has come a very important part of the way to creating a system for multilingual sites and is almost at the end of the road to that goal.

There is only one very important issue to be solved - the translation of the home page - I am currently solving this using redirect from default home page to another translated version of the home page, redirection according to the language active on the site, but it is not entirely correct because the URL of the redirected page no longer has the front page format and cannot be managed with Metatags, for example.

So I guess I'll wait for Backdrop to go the rest of the way for multilingual sites in its own way, in the core itself, without the need for an external module, and until then I'll probably start creating redirects to different translations of taxonomic terms in the same way. as for the home page.

And, to add, I've even come up with a way to translate data field labels by creating custom templates for those fields and using the t () function in these custom field labels. I will be happy if Backdrop solves and this problem in the future - translating field labels.

Not everyone knows these old Drupal techniques and especially for new users of Backdrop CMS with multilingual sites it would be much easier if they do not have to edit custom templates, do not create redirects, but translate everything necessary through the administrative panel of the system.

I am confident that these issues are being worked on and thought about. But I would like to emphasize that as Backdrop's users grow and its distribution geography expands, solving this problem becomes increasingly important for the success of the system at international "market".

**Comment by amilenkov** [comment]
Just as an example of a custom field template for translating a field label:

Instead of 

<div class="field-label"><?php print $label ?>:&nbsp;</div> 

in standard field template I add to my theme custom field--field-packing.tpl.php where this line is changed: 

<div class="field-label"><?php print t('Packing:');?>&nbsp;</div>

**Comment by amilenkov** [comment]
Something didn't work, I'm writing it not as a code:

div class="field-label"> php print t('Packing:') &nbsp; closing div

**Comment by herbdool** [comment]
@amilenkov I added backticks ( ` ) around your code so now it appears correctly

**Comment by herbdool** [comment]
The philosophy thus far regarding translation in Backdrop was to include the simple stuff in core and allow for more complex functionality in contrib. So that made it so that i18n module would need to accommodate some things which were now in core, such as simple term translation. Extending the concept of translation sets (which is currently only available for nodes) to taxonomy terms would need to be done by i18n or something.

If it had been easy to port i18n I would have championed doing that first before changing core, but as it turns out i18n relied on the Variable module which is not really compatible with the way config is done in Backdrop.

**Comment by indigoxela** [comment]
> Extending the concept of translation sets (which is currently only available for nodes) to taxonomy terms would need to be done by i18n

That's exactly what **i18n_taxonomy** does. It provides the ability to:

- Localize terms (translate via t()) or
- Create translation sets via **i18n_translation**

> So I'm worried that installing the i18n won't conflict with the new translation techniques introduced in Backdrop.

@amilenkov it shouldn't. But testing is welcome - on a dedicated test webspace, not on a live site

Regarding translatable field labels: that's what **i18n_field** provides.

> If it had been easy to port i18n...

Yep, it wasn't, for several reasons. But currently it's stalled because nobody showed any interest in months. Neither in testing nor in helping with any of the open issues.
In other words: if nobody does it, it won't happen.

**Comment by amilenkov** [comment]
@herbdool thanks for help:-)

I have a feeling that the issue with translation sets for taxonomy terms would be resolved better "something" than through i18n. 

i18n is a heavy paradigm and a very complex system. I have witnessed its development since Drupal 6 and it seems to me that in i18n the efforts of many different people and trends are intertwined. Sometimes in such cases it is easier to demolish the building and rebuild it than to repair structural defects. I'm not a specialist, I just have that feeling and I think by analogy with problems that I had to solve.

Paradigms are sometimes unpleasant because they prevent you from seeing another way of doing things.

I would be happy if an original and simpler backdrop way to do this is invented.

I have upgraded sites with Drupal 6 to Drupal 7, where i18n was once installed in Drupal 6 and then uninstalled - correctly with a complete uninstallation, not just deactivation. And then in the upgraded Drupal 7 site I found in the database tables from old Drupal 6 deinstalled i18n, which should not be there and do not serve any work, but for me this is a potential sign of possible future problems with the site.

So experience has taught me to treat i18n with great respect because of the tremendous work and effort involved in this system of modules, but I try to avoid installing it unless it is absolutely necessary for a complex multilingual site.

**Comment by amilenkov** [comment]
@indigoxela thanks, what you write is very true, I will try to find time to participate in testing the i18n.

**Comment by amilenkov** [comment]
@indigoxela

May I ask how to make i18n testing compliant with backdrop community testing standards. 

Because I'm self-taught inventor in everything I know about site development, with good understanding of HTML, CSS, Drupal (6 and 7), Backdrop and the very little PHP knowledge.

I work with a shared reseller hosting account from the largest Bulgarian hosting company. I have a lot of free space in this shared hosting and dozens of domains that I administer through CPanel's WHM. For me, making a test site means installing Backdrop on a subdomain in one of my own domains (not on customer domains) and working there.

I have worked actively with XAMPP and even now I have active XAMPP on my local computer, which I use primary for PHP exercises. I have developed some dev sites for testing with XAMPP and I know how to do this, but I prefer to develop them as subdomains of a real shared hosting accounts because I think it is safer - on the real web server the tested dev installation works with the real versions of PHP and the database and all configuration of a live Linux web server which the real site will also use. 

And besides, it's easier and faster for me. And more securely - it happened to me to reinstall my computer after a failure and everything done in XAMPP disappears after the reinstallation. On a real web server it is more secure for me there I can even at any time back up the current state with cPanel before making any changes to the test site and if they fail to return it to a previous working state. In addition, the hosting provider makes an automatic backup of the web server every night.

With what has been written above, I do not claim in any way that this is the right way. I'm not sure at all that it's all professional, but for me as a self-taught person - it works perfectly for many years.

Now, however, when I decided to try testing i18n, which should be useful not only for me, but also for you and other community members, I thought that maybe there is another better way to do it. So that every other tester can log in to the same site as a user. Something like this happened when I was working on the transliteration in Bulgarian - when creating PR, I also received a test installation of Backdrop, which, however, others also could test. At the same time, I want to have access to the root directory of the tested site and to know what is installed, to be able to work with the files, templates and code in the active tested installation.

If there is such a solution and it is not so complicated that I have to study it for a few months, I will do testing i18n n the way you recommended. If I can't - I'll do it the way I'm used to - with a subdomain of one of my sites.  

And I will probably have to post text post with screenshots about my work with i18n because from the experience I already have here, I notice that professionals do not seem to want to enter as users unfamiliar websites created on unknown platforms.

**Comment by indigoxela** [comment]
@amilenkov cool, many thanks for offering testing the module!

First of all: i18n has its [own issue queue](https://github.com/backdrop-contrib/i18n/issues). All your findings should be posted there.

> May I ask how to make i18n testing compliant with backdrop community testing standards.

Everything is helpful. You've been using i18n before with Drupal, so you'd probably use it in Backdrop the same way and report what works (and what doesn't).

You don't need to provide pull requests, it's already useful if you report. The transliteration file was an exception, because non of us would have been able to provide it.

> I'm self-taught inventor in everything I know about site development

Most of us are. :wink: 

> I prefer to develop them as subdomains of a real shared hosting account [...] with the real versions of PHP and the database and all configuration of a live Linux web server which the real site will also use.

I fully agree on that.

> ... because from the experience I already have here, I notice that professionals do not seem to want to enter as users unfamiliar websites created on unknown platforms

It's usually not necessary to provide a complete testing web, as long as you can describe the steps to reproduce something. But if that's hard, a testing web can be useful to figure out the root cause of a problem.

So, see you over there in the i18n issue queue. :wave:

**Comment by amilenkov** [comment]
@indigoxela

What you wrote encouraged me and I decided to really get actively involved in testing the i18n.

And tonight I created this site:

https://i18n.site

It's already 10:14 PM here now and I'm stopping work for tonight. 

However, my plan is to make a full test multilingual site at this URL in the coming days and weeks to try out all the important features.

This is a site with a separate standalone domain (I used the promotion these days and my domain cost less than 1.5 euros). On standalone shared hosting in my reseller hosting account.

I do not know how it will be right, but it will provide full access to the site and the hosting account, to anyone involved in this task, who is authorized by the community to work on this issue.

I repeat, I do not know if this is the right way, but I am doing what I can and according to the knowledge I have.

I am writing this here because our previous correspondence here gives context and explains what I am doing, but please if what I am doing is useful, share it in the right place - https://github.com/backdrop-contrib/i18n/issues - let the participants there know that this is not someone's random idea, but in accordance with a previous discussion and your work guidelines.
